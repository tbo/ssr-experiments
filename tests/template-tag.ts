import { render, html } from '../benchmarks/streaming-template-tag/page';
import { Readable } from 'stream';

const toString = async (stream: Readable) => {
  let buffer = '';
  stream.on('data', (data: string) => (buffer += data));
  return new Promise((resolve) => stream.on('end', () => resolve(buffer)));
};

const nextTick = () => new Promise((resolve) => setImmediate(resolve));

const expectJSXtoMatchSnapshot = async (getComponent: () => any[] | Promise<any[]>) =>
  expect(await toString(render(getComponent()))).toMatchSnapshot();

describe('Template tag', () => {
  test('Render simple template tag', async () => {
    await expectJSXtoMatchSnapshot(() => html`test`);
  });

  test('Render simple template tag with number substitution', async () => {
    await expectJSXtoMatchSnapshot(() => html`test${42}`);
  });

  test('Render async template tag ', async () => {
    await expectJSXtoMatchSnapshot(async () => html`async test`);
  });

  test('Render embedded async string', async () => {
    await expectJSXtoMatchSnapshot(() => html`${Promise.resolve('async')} test`);
  });

  test('Render embedded async number', async () => {
    await expectJSXtoMatchSnapshot(() => html`async ${Promise.resolve(42)}`);
  });

  test('Render embedded template tag', async () => {
    await expectJSXtoMatchSnapshot(() => html`another ${html`template tag`}`);
  });

  test('Render embedded list', async () => {
    await expectJSXtoMatchSnapshot(() => html`another ${['Hello', '<b>World</b>']}`);
  });
  test('Render embedded async template tag', async () => {
    await expectJSXtoMatchSnapshot(() => html`another async ${Promise.resolve(html`template tag`)}`);
  });

  test('stream template tag', async () => {
    let trigger: any;
    let output = '';
    render(html`another async ${new Promise((resolve) => (trigger = resolve))}`).on('data', (data) => (output += data));
    await nextTick();
    expect(output).toMatchSnapshot();
    trigger();
    await nextTick();
    expect(output).toMatchSnapshot();
  });
});
