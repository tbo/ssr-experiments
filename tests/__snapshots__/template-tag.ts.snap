// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template tag Render async template tag  1`] = `"async test"`;

exports[`Template tag Render embedded async number 1`] = `"async 42"`;

exports[`Template tag Render embedded async string 1`] = `"async test"`;

exports[`Template tag Render embedded async template tag 1`] = `"another async template tag"`;

exports[`Template tag Render embedded list 1`] = `"another Hello&lt;b&gt;World&lt;/b&gt;"`;

exports[`Template tag Render embedded template tag 1`] = `"another template tag"`;

exports[`Template tag Render simple template tag 1`] = `"test"`;

exports[`Template tag Render simple template tag with number substitution 1`] = `"test42"`;

exports[`Template tag stream template tag 1`] = `""`;

exports[`Template tag stream template tag 2`] = `"another async undefined"`;
